test/prop-basic/input

```
@@ -0,0 +1 @@
+:load test/prop-basic/prop-basic.disco
```

Once we fix the problem with the evaluation context for `:test`, we should add some uses of these properties here with `:test`.

---

Done. It turns out randomized testing (this approach at least) is not very good at surjectivity, though -- the inner exists has to chance upon the right value, not just once but 100 times.

---

Interesting.  So when I try `:test surjective(\x.x)` I get
```
  - No example was found: surjective (λx. x)
    Checked 100 possibilities.
```
So if I understand you correctly, you're saying that what we really want is for this to succeed and tell us it tried 100 examples for the outer forall and didn't find any counterexamples, hence it is likely to be true (clearly `holds` cannot be expected to terminate).  But it is exceedingly unlikely that the inner exists will get lucky 100 times and happen to pick the same value that was picked for the forall.

I don't really know how to do better in a principled way.  It might be a nice project for someone (e.g. a student in the future) to implement a sort of best-effort algebraic solver which could look at `exists x. x == 3` and just figure out what `x` should be rather than trying randomly (but could still fall back to random search in cases where the algebraic solver fails).

---

I don't know that it's reasonable to expect this to succeed, though it's too bad it doesn't. I could imagine a policy of trying more examples on the inner quantifier (the outer one magnifies the chance of error, so you want to drive the per-example error rate down lower), but it's probably not worth it.

(And ofc when qualified types happen the `surjective` in the standard library can just have documentation to the effect of "randomized testing won't be very effective when the domain is infinite".)

---

Yeah, agreed.